File: packages/grpr-py/src/grpr/schema.py
from __future__ import annotations

from typing import Any, Dict, Literal, TypedDict

GrprLevel = Literal["trace", "debug", "info", "warn", "error", "fatal"]

GrprSourceKind = Literal["sdk", "stdout", "stderr", "mcp"]


class GrprSource(TypedDict, total=False):
    kind: GrprSourceKind
    file: str
    line: int


class GrprEvent(TypedDict, total=False):
    id: str
    ts: str
    level: GrprLevel
    type: str
    service: str
    run_id: str
    session_id: str
    message: str
    data: Dict[str, Any]
    tags: Dict[str, str]
    trace_id: str
    span_id: str
    source: GrprSource


class GrprRedactionConfig(TypedDict):
    enabled: bool
    keys: list[str]
    patterns: list[str]


class GrprMcpConfig(TypedDict):
    max_results: int
    default_lookback_ms: int


class GrprConfig(TypedDict):
    version: int
    enabled: bool
    store_dir: str
    default_service: str
    redaction: GrprRedactionConfig
    mcp: GrprMcpConfig

File: packages/grpr-core/src/schema.ts
export type GrprLevel = "trace" | "debug" | "info" | "warn" | "error" | "fatal";

export type GrprSourceKind = "sdk" | "stdout" | "stderr" | "mcp";

export type GrprSource = {
  kind: GrprSourceKind;
  file?: string;
  line?: number;
};

export type GrprSdkConfig = {
  enabled: boolean;
  capture_stdout: boolean;
  capture_stderr: boolean;
};

export type GrprEvent = {
  id: string;
  ts: string;
  level: GrprLevel;
  type: string;
  service: string;
  run_id: string;
  session_id?: string;
  message?: string;
  data?: Record<string, unknown>;
  tags?: Record<string, string>;
  trace_id?: string;
  span_id?: string;
  source?: GrprSource;
};

export type GrprConfig = {
  version: 1;
  enabled: boolean;
  store_dir: string;
  default_service: string;
  sdk: GrprSdkConfig;
  redaction: {
    enabled: boolean;
    keys: string[];
    patterns: string[];
  };
  mcp: {
    max_results: number;
    default_lookback_ms: number;
  };
};

export type GrprSearchParams = {
  service?: string;
  session_id?: string;
  run_id?: string;
  types?: string[];
  levels?: GrprLevel[];
  contains?: string;
  since?: string;
  until?: string;
  limit?: number;
  config_path?: string;
};

export type GrprStatsParams = {
  service?: string;
  session_id?: string;
  since?: string;
  until?: string;
  group_by: "type" | "level" | "stage";
  limit?: number;
  config_path?: string;
};

export type GrprSessionsParams = {
  service?: string;
  since?: string;
  limit?: number;
  config_path?: string;
};

export type GrprTailParams = {
  service?: string;
  session_id?: string;
  run_id?: string;
  limit?: number;
  config_path?: string;
};
